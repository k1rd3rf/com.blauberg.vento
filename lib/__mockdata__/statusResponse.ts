import { Response, Parameter } from 'blaubergventojs';

const onResponse = { parameter: 1, value: new Uint8Array([1]) };

const dataEntries = [
  { parameter: 2, value: new Uint8Array([1]) },
  { parameter: 68, value: new Uint8Array([179]) },
  { parameter: 6, value: new Uint8Array([0]) },
  { parameter: 102, value: new Uint8Array([30]) },
  { parameter: 183, value: new Uint8Array([2]) },
  { parameter: 136, value: new Uint8Array([0]) },
  { parameter: 100, value: new Uint8Array([34, 20, 31, 0]) },
  { parameter: 37, value: new Uint8Array([45]) },
  { parameter: 15, value: new Uint8Array([1]) },
  { parameter: 25, value: new Uint8Array([70]) },
  { parameter: 185, value: new Uint8Array([4, 0]) },
  { parameter: 74, value: new Uint8Array([236, 4]) },
  { parameter: 7, value: new Uint8Array([0]) },
  { parameter: 131, value: new Uint8Array([0]) },
  { parameter: 11 as Parameter, value: new Uint8Array([0, 0, 0]) },
];

export const statusResponse: Response = {
  // @ts-expect-error: mock data
  packet: {
    _deviceId: 'devId',
    _password: '1234',
    _functionType: 6,
    dataEntries: [onResponse, ...dataEntries],
  },
  ip: '127.0.0.1',
};

export const offResponse: Response = {
  // @ts-expect-error: mock data
  packet: {
    _deviceId: 'devId',
    _password: '1234',
    _functionType: 6,
    dataEntries: [{ parameter: 1, value: new Uint8Array([0]) }, ...dataEntries],
  },
  ip: '127.0.0.1',
};
