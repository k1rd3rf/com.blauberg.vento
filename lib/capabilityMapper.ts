import { Response, Parameter } from 'blaubergventojs';
import { Capabilities } from './capabilities';
import { parametersToValues } from './mapModbusResponse';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const capabilityToParameterMap: Record<keyof typeof Capabilities, (values: Record<keyof typeof Parameter | string, number[] | undefined>) => any> = {
  alarm_boost: (values) => (values.BOOT_MODE?.[0] !== 0),
  alarm_filter: (values) => (values.FILTER_ALARM?.[0] === 1),
  filter_timer: (values) => [values.FILTER_TIMER?.[2], values.FILTER_TIMER?.[1], values.FILTER_TIMER?.[0]].map((n) => n?.toString().padStart(2, '0')).join(':'),
  alarm_generic: (values) => (values.READ_ALARM?.[0] !== 0),
  measure_humidity: (values) => values.CURRENT_HUMIDITY?.[0],
  measure_RPM: (values) => values.FAN1RPM?.[0],
  speedMode: (values) => values.SPEED?.[0].toString(),
  manualSpeed: (values) => ((values.MANUAL_SPEED?.[0] ?? -1) / 255) * 100,
  fan_speed: (values) => ((values.MANUAL_SPEED?.[0] ?? -1) / 255),
  operationMode: (values) => values.VENTILATION_MODE?.[0].toString(),
  timerMode: (values) => values.TIMER_MODE?.[0].toString(),
  timerMode_timer: (values) => [values['11']?.[2], values['11']?.[1], values['11']?.[0]].map((n) => n?.toString().padStart(2, '0')).join(':'),
  alarm_connectivity: (values) => !values,
  onoff: (values) => values.ON_OFF?.[0] === 1,
};

export default (result: Response) => {
  const values = parametersToValues(result);
  return Object.keys(Capabilities).map((capability) => {
    // @ts-expect-error: ok now
    const mapFunction = capabilityToParameterMap[capability];
    if (mapFunction) {
      return { [capability]: mapFunction(values) };
    }
  }).reduce((a, b) => ({ ...a, ...b }), {});
};
