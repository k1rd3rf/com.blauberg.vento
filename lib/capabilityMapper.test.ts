import capabilityMapper from './capabilityMapper';
import { statusResponse } from './__mockdata__/statusResponse';

describe('capabilityMapper', () => {
  it('maps packet to capability response', () => {
    expect(capabilityMapper(statusResponse)).toMatchSnapshot();
  });
});
