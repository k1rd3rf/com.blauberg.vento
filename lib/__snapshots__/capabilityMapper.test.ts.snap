// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`capabilityMapper maps packet to capability response 1`] = `
{
  "alarm_boost": false,
  "alarm_filter": false,
  "alarm_generic": false,
  "boost_delay": 30,
  "fan_speed": 0.7019607843137254,
  "filter_timer": "31:20:34",
  "humidity_sensor": true,
  "humidity_threshold": 70,
  "manualSpeed": 70.19607843137254,
  "measure_RPM": 236,
  "measure_humidity": 45,
  "onoff": true,
  "operationMode": "2",
  "speedMode": "1",
  "timerMode": "0",
  "timerMode_timer": "00:00:00",
  "unit_type": "Vento Expert Duo A30-1 W V.2",
}
`;
